### ISSUE TITLE
go_library should fail when a direct dependency is missing

### ISSUE THREAD

**Issue by jayconrod** [bazelbuild/rules_go#566]
Suppose we have this BUILD file:

```
load("@io_bazel_rules_go//go:def.bzl", "go_library")

go_library(
    name = "go_default_library",
    srcs = ["lib.go"],
    deps = [":a"],
)

go_library(
    name = "a",
    srcs = ["a.go"],
    deps = [":b"],
)

go_library(
    name = "b",
    srcs = ["b.go"],
)
```

If lib.go imports both "a" and "b", Bazel should report an error because the dependency on "b" wasn't explicitly specified in `deps`. Currently, no error is reported. Transitive dependencies must be available to the compiler, since we need type information, so the compiler won't be able to check this.

**Comment by pmbethe09** [comment]
as I understand it, bazel dumps the direct dependencies into a .jdeps file (as read by the unused_deps tool), and then checks
(direct intersect declared) == direct
and errors if not so.

We could run an action that does something similar?

**Comment by jayconrod** [comment]
Unfortunately, I think .jdeps files are only produced for Java rules.

I'm thinking the solution to this will be to parse the imports of the filtered .go files (ignoring files that fail the filter) and checking those imports are provided by direct deps. We'll need to pipe the list of direct deps in from the rules, but it shouldn't be too complicated.
